module-system/sov-state/src/vec.rs

```diff
@@ -0,0 +1,201 @@
+use std::marker::PhantomData;
+
+use borsh::{BorshDeserialize, BorshSerialize};
+use thiserror::Error;
+
+use crate::codec::{BorshCodec, StateKeyCodec, StateValueCodec};
+use crate::{Prefix, Storage, WorkingSet};
+
+#[derive(Debug, PartialEq, Eq, Clone, BorshDeserialize, BorshSerialize)]
+pub struct StateVec<V, C = BorshCodec> {
+    _phantom: PhantomData<V>,
+    codec: C,
+    prefix: Prefix,
+}
+
+/// Error type for `StateVec` get method.
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("Index out of bounds for index: {0}")]
+    IndexOutOfBounds(usize),
+    #[error("Value not found for prefix: {0} and index: {1}")]
+    MissingValue(Prefix, usize),
+}
+
+impl<V> StateVec<V>
+where
+    BorshCodec: StateValueCodec<V>,
+{
+    /// Crates a new [`StateVec`] with the given prefix and the default
+    /// [`StateCodec`] (i.e. [`BorshCodec`]).
+    pub fn new(prefix: Prefix) -> Self {
+        Self {
+            _phantom: PhantomData,
+            codec: BorshCodec,
+            prefix,
+        }
+    }
+}
+
+impl<V, C> StateVec<V, C>
+where
+    C: StateValueCodec<V>,
+    C: StateValueCodec<usize>,
+{
+    /// Creates a new [`StateVec`] with the given prefix and codec.
+    pub fn with_codec(prefix: Prefix, codec: C) -> Self {
+        Self {
+            _phantom: PhantomData,
+            codec,
+            prefix,
+        }
+    }
+
+    /// Returns the prefix used when this [`StateVec`] was created.
+    pub fn prefix(&self) -> &Prefix {
+        &self.prefix
+    }
+
+    fn internal_codec(&self) -> IndexCodec<C> {
+        IndexCodec::new(&self.codec)
+    }
+
+    /// Sets a value in the [`StateVec`].
+    /// If the index is out of bounds, returns an error.
+    /// To push a value to the end of the StateVec, use [`StateVec::push`].
+    pub fn set<S: Storage>(&self, index: usize, value: &V, working_set: &mut WorkingSet<S>) -> Result<(), Error> {
+        let len = self.len(working_set);
+
```

Comment:
should we call this replace? You cannot arbitrarily set index, it can only used when changing existing value?

Reply:
I think of this like a regular array

```
arr = [1,2,3]

arr[2] = 5 // this works
arr[3] = 5 // this wouldn't and would throw index out of bounds on many languages
```

so imo the name set makes sense